src/debugAdapter/goDebug.ts

```diff
@@ -479,71 +487,93 @@ class Delve {
+	public callPromise<T>(command: string, args: any[]): Thenable<T> {
 		return new Promise<T>((resolve, reject) => {
 			this.connection.then(conn => {
-				conn.call<T>('RPCServer.' + command, args, (err, res) => {
-					if (err) return reject(err);
-					resolve(res);
+				conn.call<T>(`RPCServer.${command}`, args, (err, res) => {
+					return err ? reject(err) : resolve(res);
 				});
 			}, err => {
 				reject(err);
 			});
 		});
 	}
 
-	close(): Thenable<void> {
+	/**
+	 * To close the debugging session we do the following; If its a local process,
+	 * we try to halt and detach within a timeout otherwise we kill the debugger. If its a remote
+	 * process we try to halt and restart the remote target process running under the debugger.
+	 */
+	public close(): Thenable<void> {
 		if (this.noDebug) {
 			// delve isn't running so no need to halt
 			return Promise.resolve();
 		}
 		log('HaltRequest');
 
-		return new Promise(resolve => {
-			let timeoutToken: NodeJS.Timer;
-			if (this.debugProcess) {
-				timeoutToken = setTimeout(() => {
-					log('Killing debug process manually as we could not halt and detach delve in time');
-					killTree(this.debugProcess.pid);
-					resolve();
-				}, 1000);
-			}
+		return new Promise(async resolve => {
+			const timeoutToken: NodeJS.Timer = !this.isRemoteDebugging() && setTimeout(() => {
+				log('Killing debug process manually as we could not halt delve in time');
+				killTree(this.debugProcess.pid);
+				this.ensureDebugeeExecutableIsRemoved();
+				resolve();
+			}, 1000);
 
-			this.callPromise('Command', [{ name: 'halt' }]).then(() => {
-				if (timeoutToken) {
-					clearTimeout(timeoutToken);
-				}
+			let errMsg;
+			try {
+				await this.callPromise('Command', [{ name: 'halt' }]);
+			} catch (err) {
 				log('HaltResponse');
-				if (!this.debugProcess) {
-					log('RestartRequest');
-					return this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }])
-						.then(null, err => {
-							log('RestartResponse');
-							logError(`Failed to restart - ${(err || '').toString()}`);
-						})
-						.then(() => resolve());
-				} else {
-					log('DetachRequest');
-					return this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }])
-						.then(null, err => {
-							log('DetachResponse');
-							logError(`Killing debug process manually as we failed to detach - ${(err || '').toString()}`);
-							killTree(this.debugProcess.pid);
-						})
-						.then(() => resolve());
-				}
-			}, err => {
-				const errMsg = err ? err.toString() : '';
-				log('Failed to halt - ' + errMsg.toString());
-				if (errMsg.endsWith('has exited with status 0')) {
-					if (timeoutToken) {
-						clearTimeout(timeoutToken);
-					}
-					return resolve();
+				errMsg = err ? err.toString() : '';
+				log(`Failed to halt - ${errMsg}`);
+			}
+			clearTimeout(timeoutToken);
+
+			if (this.shouldDetach(errMsg)) {
```

Comment:
`shouldDetach()` and `targetHasExited()` are not being used anywhere else. Can we have them as `const` before this `if` instead of functions?